pallets/energy/src/lib.rs

```diff
@@ -0,0 +1,203 @@
+        ValueQuery,
+    >;
+
+    #[pallet::call]
+    impl<T: Config> Pallet<T> {
+        /// Generate energy for a target account by burning balance from the caller.
+        #[pallet::weight(10_000)]
+        pub fn generate(
+            origin: OriginFor<T>,
+            target: <T::Lookup as StaticLookup>::Source,
+            amount: BalanceOf<T>,
+        ) -> DispatchResult {
+            let caller = ensure_signed(origin)?;
+            let target = T::Lookup::lookup(target)?;
+
+            ensure!(T::Currency::can_slash(&caller, amount), Error::<T>::NotEnoughBalance);
+            let _ = T::Currency::slash(&caller, amount);
+
+            Self::generate_energy(&target, amount)?;
+
+            Self::deposit_event(Event::EnergyGenerated {
+                generator: caller,
+                receiver: target,
+                burnt_balance: amount,
+                generated_energy: amount,
+            });
+
+            Ok(())
+        }
+    }
+
+    impl<T: Config> Pallet<T> {
+        fn generate_energy(target: &T::AccountId, amount: BalanceOf<T>) -> DispatchResult {
+            TotalEnergy::<T>::mutate(|total| -> Result<(), ArithmeticError> {
+                *total = total.checked_add(&amount).ok_or(ArithmeticError::Overflow)?;
+                Ok(())
+            })?;
+            EnergyPerAccount::<T>::mutate(target, |energy| -> Result<(), ArithmeticError> {
+                *energy = energy.checked_add(&amount).ok_or(ArithmeticError::Overflow)?;
+                Ok(())
+            })?;
+            Ok(())
+        }
+
+        fn consume_energy(target: &T::AccountId, amount: BalanceOf<T>) -> DispatchResult {
+            EnergyPerAccount::<T>::mutate(target, |energy| -> Result<(), ArithmeticError> {
+                *energy = energy.checked_sub(&amount).ok_or(ArithmeticError::Underflow)?;
+                Ok(())
+            })?;
+            TotalEnergy::<T>::mutate(|total| -> Result<(), ArithmeticError> {
+                *total = total.checked_sub(&amount).ok_or(ArithmeticError::Underflow)?;
+                Ok(())
+            })?;
+            Ok(())
+        }
+    }
+
+
+    pub enum LiquidityInfo<T: Config> {
+        /// Nothing have been paid.
+        Nothing,
+        /// Transaction have been paid using energy.
+        Energy(BalanceOf<T>),
+        /// Transaction have been paid using the fallback method.
+        Fallback(<T::FallbackOnChargeTransaction as OnChargeTransaction<T>>::LiquidityInfo),
+    }
+
+    impl<T: Config> Default for LiquidityInfo<T> {
+        fn default() -> Self {
+            LiquidityInfo::Nothing
+        }
+    }
+
+    impl<T: Config> OnChargeTransaction<T> for Pallet<T> {
+        type Balance = BalanceOf<T>;
+        type LiquidityInfo = LiquidityInfo<T>;
+
+        fn withdraw_fee(
+            who: &T::AccountId,
+            call: &T::Call,
+            dispatch_info: &DispatchInfoOf<T::Call>,
+            fee: Self::Balance,
+            tip: Self::Balance,
```

Comment:
Don't forget to finalize thoughts on paying tips.
A little remark: we shouldn't be able to pay tips with energy.